src/canvas.js

```diff
@@ -1,22 +1,39 @@
 import { getDataKeySync } from './utils.js'
 import Seedrandom from 'seedrandom'
 
-export function computeOffScreenCanvas (canvas, domainKey, sessionKey, getImageDataProxy) {
-    const ctx = canvas.getContext('2d')
-    // We *always* compute the random pixels on the complete pixel set, then pass back the subset later
-    let imageData = getImageDataProxy._native.apply(ctx, [0, 0, canvas.width, canvas.height])
-    imageData = modifyPixelData(imageData, sessionKey, domainKey, canvas.width)
+export function computeOffScreenCanvas (canvas, domainKey, sessionKey, getImageDataProxy, ctx) {
+    if (!ctx) {
+        ctx = canvas.getContext('2d')
+    }
 
     // Make a off-screen canvas and put the data there
     const offScreenCanvas = document.createElement('canvas')
     offScreenCanvas.width = canvas.width
     offScreenCanvas.height = canvas.height
     const offScreenCtx = offScreenCanvas.getContext('2d')
+
+    offScreenCtx.drawImage(canvas, 0, 0)
+
+    // We *always* compute the random pixels on the complete pixel set, then pass back the subset later
+    let imageData = getImageDataProxy._native.apply(offScreenCtx, [0, 0, canvas.width, canvas.height])
+    imageData = modifyPixelData(imageData, sessionKey, domainKey, canvas.width)
+
+    clearCanvas(offScreenCtx)
+
     offScreenCtx.putImageData(imageData, 0, 0)
 
     return { offScreenCanvas, offScreenCtx }
 }
 
+function clearCanvas (canvasContext) {
```

Comment:
> Why move away from the previous design which just computed the modified pixels directly from the input canvas?

We have to rasterize 3d canvases into 2d ones so that we can call getImageData on it and do our algorithm. I couldn't think of a way to easily make it work over toDataURL which is all the 3d context support.

Clearing the pixels just allows the element reuse and ensures we're not applying pixels on top of each other.

Reply:
Got it. That makes sense!